### DISCUSSION TITLE
How to Manually Focus a Specific Item in FlatList Using Remote Control in React Native TVOS?

### DISCUSSION THREAD

**Discussion by saadlatif75** [react-native-tvos/react-native-tvos#837] [Q&A] [answered]
Hi everyone,

I'm working on a React Native TVOS app, and I need to manually focus a specific item in a FlatList when using a remote control. For instance, based on some user interaction or state, I want to programmatically move the focus to a particular item in the list.

Here's what I've tried so far:

Using ref to scroll to the desired index with scrollToIndex. It scrolls correctly, but the item doesn’t get focused automatically.
Setting the focus manually using TVFocusGuideView or wrapping items with TouchableHighlight/Pressable and trying to call .focus().
Is there a way to directly programmatically focus an item in a FlatList while navigating with a remote control? Any suggestions or workarounds would be greatly appreciated!

Thanks in advance! 😊

**Comment by douglowder** [comment]
See https://github.com/react-native-tvos/react-native-tvos/discussions/664

Questions for you:
- Are you testing on Apple TV or Android TV?
- What version of RNTV are you using?
- Could you share sample code showing how you are using `scrollToIndex`?

@yusufyildirim

**Reply by saadlatif75** [reply]
@douglowder @yusufyildirim 
 1.i'm testing on Android TV  
 2. "react-native": "npm:react-native-tvos@0.73.6-1",
 3.
 const listRef = useRef(null); // Create a reference to the FlashList
  const [focusedItemIndex, setFocusedItemIndex] = useState(null);
  const handleUpArrowPress = () => {
    if (focusedItemIndex === 0 && listRef.current) {
      listRef?.current?.scrollToIndex({
        index: data?.length - 1,
        animated: true,
      });
    }
  };
   **const handleFocus = item => {
    setFocusedItemIndex(data?.findIndex(i => i?.id === item?.id));
    // toggleDrawer();
  };**
  ```
 <FlashList
          data={data}
          ref={listRef}
          renderItem={renderSubMenu}
          keyExtractor={keyExtractor}
          ListHeaderComponent={headerVisiblity}
          estimatedItemSize={56}
        />
```

**Reply by douglowder** [reply]
Does this problem happen with FlatList from this repo? FlashList is a 3rd party component.

**Reply by saadlatif75** [reply]
> Does this problem happen with FlatList from this repo? FlashList is a 3rd party component.

It's not specific to FlashList, as i encountered similar issues with the
native FlatList component from React Native.

**Comment by yusufyildirim** [comment, marked as answer]
Hey, there's no built-in way to do that at the moment. See my feature proposal for that here: https://github.com/react-native-tvos/react-native-tvos/discussions/665

I see that I didn't explain the problem in the proposal that well, I should probably update it. In a nutshell, what you want to achieve requires a bit acrobatic 😄  You should follow this guideline:

Don't forget that FlatList is virtualized. The item you want to focus on should be intact in the UI tree. So, your virtualization configuration should be proper. Even if you make sure the component you want to focus on is *rendered*, it doesn't mean it's attached to the UI tree, don't forget that. There's this property called [removeClippedSubviews](https://reactnative.dev/docs/virtualizedlist#removeclippedsubviews) (which is enabled on Android by default) that makes sure everything outside the viewport is *detached* from the UI tree.

Long story short, to make your life simpler (not necessarily easier), you should make sure the component you want to focus on *is* visible on the screen before calling focus action on it. To achieve that;

Programatically scroll your list to your element's position. Don't forget the items are virtualized and the rendering is async. Meaning, your item might not be there for a split second even though you scrolled to the correct position. So, scrolling to the position is not a guarantee. You might need to use requestAnimationFrame or something like that depending on your usecase to give React enough time to render your items. Do not use timers, it's never a good idea, just saying 😄 

If you want to focus on a specific item/region every time, check out the README and look for `additionalRenderRegions` property (at the very bottom). You can make some items exempt from virtualization, thus  simplifying your life.

If you align the planets perfectly, you can call the `requestTVFocus`, focus on the item and enjoy the view 😄

Sorry for the sarcasm 😄 I also deal with this complexity at work every day. I will provide a solution for that one day, just don't know when. Good luck!

**Comment by douglowder** [comment]
@saadlatif75 I have migrated this issue to a discussion in the Q&A section.

**Comment by saadlatif75** [comment]
It's not specific to FlashList, as i encountered similar issues with the
native FlatList component from React Native.
